pkg/capabilities/consensus/ocr3/requests/handler.go

```diff
@@ -0,0 +1,146 @@
+package requests
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+type responseCacheEntry struct {
+	response  *Response
+	entryTime time.Time
+}
+
+type Handler struct {
+	services.StateMachine
+
+	lggr logger.Logger
+
+	mux             sync.Mutex
+	store           *Store
+	responseCache   map[string]*responseCacheEntry
+	cacheExpiryTime time.Duration
+
+	stopCh services.StopChan
+	wg     sync.WaitGroup
+}
+
+func NewHandler(lggr logger.Logger, s *Store, responseExpiryTime time.Duration) *Handler {
+	return &Handler{
+		lggr:            lggr,
+		store:           s,
+		responseCache:   map[string]*responseCacheEntry{},
+		cacheExpiryTime: responseExpiryTime,
+		stopCh:          make(services.StopChan),
+	}
+}
+
+func (h *Handler) SendResponse(resp *Response) {
+	req := h.getRequestForResponse(resp)
+	if req != nil {
+		req.sendResponse(resp)
+	}
+}
+
+func (h *Handler) SendRequest(r *Request) error {
+	resp, err := h.addRequest(r)
+	if err != nil {
+		return fmt.Errorf("failed to add request: %w", err)
+	}
+
+	if resp != nil {
+		r.sendResponse(resp)
```

Comment:
Is there a deadlock risk if we send response immediately from the same goroutine? Maybe @cedric-cordenier can confirm?

Reply:
Yes it is possible -- all responses should go via the transmitCh, which also serialises the responses